### ISSUE TITLE
validateQuoteMarks

### ISSUE THREAD

**Issue by hzoo** [jscs-dev/babel-jscs#1]
Moving issue from old repo.

Fails jscs tests / doesn't validate correctly. @deezahyn

**Comment by hzoo** [comment]
`'a'` is parsed as `"a"`.

It seems that

``` js
var parse = require("babel-core").parse;
...
ast = parse(code, opts);

// token
{ type: 'String',
  value: 'a',
  start: 8,
  end: 11,
  loc: { start: { line: 1, column: 8 }, end: { line: 1, column: 11 } },
  range: [ 8, 11 ] }
```

Only has the value in the tokens so it won't have quotes.

And then I think `JSON.stringify` always adds double quotes?

``` js
} else if (type === tt.string) {
    token.type = "String";
    token.value = JSON.stringify(token.value);
  } 
```

However it does have the quotes in the ast itself (`Literal`, node.raw)

JSCS is using the string token and checking `token.value`. Don't know if we can iterate through `Literal` instead, which `iterateNodes`? @mrjoelkemp

``` js
file.iterateTokensByType('String', function(token) {
            var str = token.value;
...
```

https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/validate-quote-marks.js

basically ast.tree has the quote marks but ast.tokens doesn't.

**Comment by hzoo** [comment]
Looks like we have to figure out how to add the correct quotes to ast.tokens in babel-jscs or change the rule in jscs itself to use ast.tree and iterate through nodes instead of tokens

**Comment by hzoo** [comment]
Ok I guess we will fix this in the jscs rule by doing the slice on start/end? @mdevils

**Comment by mdevils** [comment]
Yeah, I think it should completely fix issues.

**Comment by mdevils** [comment]
https://github.com/jscs-dev/babel-jscs/blob/20a9d2749434c8a5f72b5a79d42320394cc7a5e2/acorn-to-esprima.js#L54

I believe we should also use `slice` for `RegExp`, because `toString` for `RegExp` is way more expensive.

**Comment by hzoo** [comment]
Oh ok you mean use `token.value = source.slice(token.range[0], token.range[1]);` for all 3 then (string, numeric, regexp)?

**Comment by mdevils** [comment]
Yes, exactly.

**Comment by okonet** [comment]
This is awesome! Thanks!!! 👏💖
